### DISCUSSION TITLE
Executing tests on a spawned thread

### DISCUSSION THREAD

**Discussion by niclo** [cucumber-rs/cucumber#284] [Q&A] [answered]
In the book it states tests can be executed asynchronously as such:

```rust
#[tokio::main]
async fn main() {
    AnimalWorld::run("tests/features/book/quickstart").await;
}
```

However in my use case, that being a web server which receives a request to execute a cucumber test, I would like to run said test on a separate thread, allowing the handler to exit and once the test has been executed, the result is saved to a database. Naively I went with something like this:

```rust
tokio::spawn(async {
    AnimalWorld::cucumber().fail_fast().run("tests/features/book/quickstart").await;
});
```

However this does not work because future is not `Send`. I looked into a potential [solution from tokio](https://docs.rs/tokio/latest/tokio/task/fn.spawn.html#using-send-values-from-a-task) for this problem, however I do not think it would work with what I am trying to do.

Therefore I was wondering if the cucumber-rs team might have already encountered this particular use case and could kindly point me in the right direction. Thank you :)

**Comment by ilslv** [comment]
@niclo I'm not quite sure I understand the use-case here. 

> I would like to run said test on a separate thread

`tokio::spawn` doesn't guarantee that `Future` will be executed on a separate thread

> allowing the handler to exit and once the test has been executed, the result is saved to a database.

Can you elaborate?

**Reply by niclo** [reply]
Thank you for your fast response!

The http server has access to a registry of devices. Currently a user can send a request to manipulate certain characteristics of a device, for instance, updating the firmware of a given device. I would like to expand upon the server, by allowing users to send a request to run a specific test against a given device. That is why I need to call, e.g., `AnimalWorld::run()` outside of `main()`. The result of the test is then saved to the database, so that the user can send another request to see whether or not the test was successful.

Another solution I am currently looking into is using [std::process::Command](https://doc.rust-lang.org/std/process/struct.Command.html) to execute the test on a separate process instead of a separate thread.

Essentially I would just like to know if some form of this

```rust
tokio::spawn(async {
    AnimalWorld::run("tests/features/book/quickstart").await;
});
```

is possible or not. Thank you very much :)

**Reply by ilslv** [reply, marked as answer]
@niclo in case your tests do not contain significant blocking logic, there is no need to run tests in a separate thread. Just use `.await`. In case there is blocking code, you should wrap it in a [`tokio::spawn_blocking`](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html). 

Alternatively you can spawn a thread, [spin up a new `tokio` runtime and then block on `World::run()`](https://docs.rs/tokio/latest/tokio/runtime/index.html#usage).

**Reply by niclo** [reply]
Alright thank you 🙏🏼
